### PR TITLE
Configurable management group hierarchy

### PR DESCRIPTION
**Pull request by skeeler**
## Overview/Summary
Implement a configurable management group hierarchy that allows specifying both `id` and `name` attributes for each management group and maintains full backward compatibility with previous environment configuration files and existing code base.

## This PR fixes/adds/changes/removes

1. Implement configurable management group hierarchy with full backward compatibility (config & code). #153 
2. Allow `name` to be set (new for all MGs). #167 
3. Put environment guard on the `roles-ci` pipeline. #185 
4. Address Landing Zones management groups design. #77 

### Breaking Changes

None

## Testing Evidence

Tested the following scenarios across 6 core pipelines (nva for networking, not azfw)
- Only legacy vars present.
- New `var-managementgroup-hierarchy` config present, and **with** hard coded MG vars (`var-parentManagementGroupId` and `var-topLevelManagementGroupName`) present.
- New `var-managementgroup-hierarchy` config present, and **without** hard coded MG vars (`var-parentManagementGroupId` and `var-topLevelManagementGroupName`) present.

Please provide any testing evidence to show that your Pull Request works/fixes as described and planned (include screenshots, if appropriate).

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/azure/CanadaPubSecALZ/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/azure/CanadaPubSecALZ/issues)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/azure/CanadaPubSecALZ/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated relevant and associated documentation.

### CODE REVIEW

File: docs/onboarding/azure-devops-pipelines.md [hunk 1]
```diff
@@ -171,39 +171,83 @@ Instructions:
...
         var-parentManagementGroupId: abcddfdb-bef5-46d9-99cf-ed67dabc8783
         var-topLevelManagementGroupName: pubsec
-
+    ...
     ```
 
+    In CanadaPubSecALZ v0.8.0 or earlier, the only portion of the management group hierarchy that is configurable is the top level management group id (specified using the `topLevelManagementGroupName` attribute shown above). For these versions, observe the following guidance:
+
+    * The management group name is not configurable, it is automatically set ot the same value provided for the id.
+    * The management group hierarchy is hard-coded in the `management-groups/structure.bicep` source file. You will need to modify this file (add/edit/remove bicep code) if you want to change the management group hierarchy.
```
Comment (SenthuranSivananthan) [inline docs/onboarding/azure-devops-pipelines.md:249]: Can you link it to the source file?

File: .pipelines/templates/steps/load-variables.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# Load Variables
+- task: Bash@3
+  displayName: Load Variables
+  inputs:
+    targetType: inline
+    script: |
+      $(var-bashPreInjectScript)
+
+      ### -------------------------------------------------------------------------------------------------
+      ### Create the legacy 'var-parentManagementGroupId' and 'var-topLevelManagementGroupName'
```
Comment (SenthuranSivananthan) [inline .pipelines/templates/steps/load-variables.yml:21]: These two parameters are not legacy since they are still required for pipelines to function.  The automatic approach is to query the hierarchy and set there value.

Therefore, I wouldn't call them legacy.  We will keep these parameters indefinitely.

File: .pipelines/templates/steps/load-variables.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      ### variables (used in YAML pipeilne definitions) based on information extracted from the
+      ### newer 'var-managementgroup-hierarchy' variable, if it is defined.
+      ### -------------------------------------------------------------------------------------------------
+      hierarchyJson='$(var-managementgroup-hierarchy)'
+      if [[ -n $hierarchyJson ]]; then
+        echo "The new 'var-managementgroup-hierarchy' variable IS defined."
+        echo "REMAPPING legacy 'var-parentManagementGroupId' and 'var-topLevelManagementGroupName' variables."
```
Comment (SenthuranSivananthan) [inline .pipelines/templates/steps/load-variables.yml:28]: I would remove the reference to legacy

File: docs/onboarding/azure-devops-pipelines.md [hunk 4]
```diff
@@ -171,39 +171,83 @@ Instructions:
...
 
-    **Sample environment YAML**
+    **Sample environment YAML (v0.9.0 or later)**
 
-    ```yaml
+    ```yml
     variables:
+    
+      # Management Groups
+      var-managementgroup-hierarchy: >
```
Comment (SenthuranSivananthan) [inline docs/onboarding/azure-devops-pipelines.md:202]: The hierarchy should be a full example but it looks incomplete.

File: docs/onboarding/azure-devops-pipelines.md [hunk 5]
```diff
@@ -171,39 +171,83 @@ Instructions:
...
+      "id": "ManagementGroupIdentifier_NoSpaces",
+      "children": []
+    }
+    ```
 
+    Each `"children": []` element is an array of the same object type, and may contain zero or more child management group object definitions, which in turn may contain their own arrays of zero or more child management group object definitions.
+
+    Observe the following guidance when configuring your management group hierarchy:
+
+      * Specify the `id` and `name` of your existing Azure AD tenant for the topmost management group definition. The `id` attribute for this element is mapped into the legacy `var-parentManagementGroupId` pipeline variable for backward compatibility.
```
Comment (SenthuranSivananthan) [inline docs/onboarding/azure-devops-pipelines.md:229]: I would remove `legacy` reference

File: docs/onboarding/azure-devops-pipelines.md [hunk 6]
```diff
@@ -171,39 +171,83 @@ Instructions:
...
+      * Specify only 1 child management group definition for the topmost management group definition. You can specify more than 1 child of the topmost management group definition, but it is the first child of the topmost level that will be considered the root of of your management group hierarchy, and is the scope that the `policy-ci` pipeline will use to deploy built-in and custom policies. The `id` attribute for this element is mapped into the legacy `var-topLevelManagementGroupName` pipeline variable for backward compatibility.
```
Comment (SenthuranSivananthan) [inline docs/onboarding/azure-devops-pipelines.md:230]: i would remove the `legacy` reference

File: docs/onboarding/azure-devops-pipelines.md [hunk 7]
```diff
@@ -171,39 +171,83 @@ Instructions:
...
+      * The `id` attribute for management group elements can only be an ASCII letter, digit, `-`, `_`, `(`, `)`, `.` and cannot end with a period. In the sample environment configuration file (`CanadaESLZ-main.yml`), we illustrate a convention that prepends the id of the parent management group to the id of each child management group. This is an example only and not a requirement. You are welcome to choose any management group id convention that best suits your needs.
+      * If you are using **CanadaPubSecALZ v0.9.0 or later** and **do not** include a `var-managementgroup-hierarchy` variable setting in your configuration, it will fallback to using the legacy pipeline variables `var-parentManagementGroupId` and `var-topLevelManagementGroupName`. This is to ensure backward compatibility, enabling newer versions of the code to run with older environment configurations.
```
Comment (SenthuranSivananthan) [inline docs/onboarding/azure-devops-pipelines.md:232]: Instead of referencing the parameters as `legacy` pipeline variables, what if we say that it will fall back to hardcoded values?

File: docs/onboarding/azure-devops-pipelines.md [hunk 8]
```diff
@@ -171,39 +171,83 @@ Instructions:
...
+      * If you are using **CanadaPubSecALZ v0.9.0** or later and **do** include a `var-managementgroup-hierarchy` variable setting in your configuration, it will override any legacy pipeline variables `var-parentManagementGroupId` and `var-topLevelManagementGroupName` also present.
```
Comment (SenthuranSivananthan) [inline docs/onboarding/azure-devops-pipelines.md:233]: Let's remove `legacy`

File: docs/onboarding/azure-devops-pipelines.md [hunk 9]
```diff
@@ -267,23 +311,25 @@ This role assignment is used to grant users access to the logging subscription b
 
 1. Edit `./config/variables/<devops-org-name>-<branch-name>.yml` in Git.  This configuration file was created in Step 3.
 
-    * Update **var-logging-managementGroupId** with the logging management group.  This is based on the prefix defined in `var-topLevelManagementGroupName`.  For example, if `var-topLevelManagementGroupName` is set to `contoso`, then `var-logging-managementGroupId` will be `contosoPlatformManagement`.
-    
+    * Update **var-logging-managementGroupId** with the logging management group:
+      * For **CanadaPubSecALZ v0.9.0 or later**, this is based on the `id` attribute of the first child element defined in your `var-managementgroup-hierarchy` configuration value. In JSON notation, this is equivalent to: `.children[0].id`.
```
Comment (SenthuranSivananthan) [inline docs/onboarding/azure-devops-pipelines.md:315]: This wouldn't be the case when we have the hierarchy.  Let's rephrase it to say that with v0.9.0 or later, this will be the management id that represents the Logging Management Group in the defined hierarchy.

File: docs/onboarding/azure-devops-pipelines.md [hunk 10]
```diff
@@ -421,22 +467,24 @@ In order to configure audit stream for Azure Monitor, identify the following inf
     is optional since the deployment will be based on the pipeline you configure anyway. So if you configured the **platform-connectivity-hub-azfw-policy-ci** and the **platform-connectivity-hub-azfw-ci** pipelines,
     only the _Azure Firewall_ option will be deployed, not the _NVA_ option.
 
-    * Update **var-hubnetwork-managementGroupId** with the networking management group.  This is based on the prefix defined in `var-topLevelManagementGroupName`.  For example, if `var-topLevelManagementGroupName` is set to `contoso`, then `var-hubnetwork-managementGroupId` will be `contosoPlatformConnectivity`.
-    
+    * Update **var-hubnetwork-managementGroupId** with the networking management group:
+        * For **CanadaPubSecALZ v0.9.0 or later**, this is based on the `id` attribute of the first child element defined in your `var-managementgroup-hierarchy` configuration value. In JSON notation, this is equivalent to: `.children[0].id`.
```
Comment (SenthuranSivananthan) [inline docs/onboarding/azure-devops-pipelines.md:471]: This wouldn't be the case when we have the hierarchy.  Let's rephrase it to say that with v0.9.0 or later, this will be the management id that represents the Networking Management Group in the defined hierarchy.

File: docs/onboarding/azure-devops-pipelines.md (hunk 1 above)
Comment (skeeler) [reply docs/onboarding/azure-devops-pipelines.md]: Done

File: .pipelines/templates/steps/load-variables.yml (hunk 2 above)
Comment (skeeler) [reply .pipelines/templates/steps/load-variables.yml]: Removed 'legacy' from this step (3 occurrences)

File: .pipelines/templates/steps/load-variables.yml (hunk 3 above)
Comment (skeeler) [reply .pipelines/templates/steps/load-variables.yml]: Done

File: docs/onboarding/azure-devops-pipelines.md (hunk 5 above)
Comment (skeeler) [reply docs/onboarding/azure-devops-pipelines.md]: Done

File: docs/onboarding/azure-devops-pipelines.md (hunk 6 above)
Comment (skeeler) [reply docs/onboarding/azure-devops-pipelines.md]: Done

File: docs/onboarding/azure-devops-pipelines.md (hunk 7 above)
Comment (skeeler) [reply docs/onboarding/azure-devops-pipelines.md]: Removed legacy

File: docs/onboarding/azure-devops-pipelines.md (hunk 9 above)
Comment (skeeler) [reply docs/onboarding/azure-devops-pipelines.md]: Good catch; I was a little too quick with copy/paste from higher up in the document. Fixed.

File: docs/onboarding/azure-devops-pipelines.md (hunk 10 above)
Comment (skeeler) [reply docs/onboarding/azure-devops-pipelines.md]: Good catch; I was a little too quick with copy/paste from higher up in the document. Fixed.

File: docs/onboarding/azure-devops-pipelines.md (hunk 8 above)
Comment (skeeler) [reply docs/onboarding/azure-devops-pipelines.md]: Done

**Review by SenthuranSivananthan — APPROVED** [review]
Thank you @skeeler.  Great PR.
